src/lib.rs

```diff
@@ -408,12 +416,33 @@ pub trait ConstraintSystem<E: ScalarEngine>: Sized {
 
         Namespace(self.get_root(), PhantomData)
     }
+
+    /// Most implementations of ConstraintSystem are not 'extensible': they won't implement a specialized
+    /// version of `extend` and should therefore also keep the default implementation of `is_extensible`
+    /// so callers which optionally make use of `extend` can know to avoid relying on it when unimplemented.
+    fn is_extensible() -> bool {
+        false
+    }
+
+    /// Extend concatenates two constraint systems, destructively modifying the receiver, whose
+    /// inputs, allocated variables, and constraints will precede those of the `other` constraint system.
+    /// The primary use case for this is parallel synthesis of circuits which can be decomposed into
+    /// entirely independent sub-circuits. Each can be synthesized in its own thread, then the
+    /// original `ConstraintSystem` can be extended with each, in the same order they would have
+    /// been synthesized sequentially.
+    fn extend(&mut self, _other: &mut Self) {
+        unimplemented!(
+            "ConstraintSystem::extend must be implemented for types implementing ConstraintSystem"
+        );
+    }
 }
 
 /// This is a "namespaced" constraint system which borrows a constraint system (pushing
 /// a namespace context) and, when dropped, pops out of the namespace context.
 pub struct Namespace<'a, E: ScalarEngine, CS: ConstraintSystem<E>>(&'a mut CS, PhantomData<E>);
 
+unsafe impl<'a, E: ScalarEngine, CS: ConstraintSystem<E>> Send for Namespace<'a, E, CS> {}
```

Comment:
This is not sound, `Send` is automatically derived when all generic args are send.

Reply:
Hmmm.

> Types that aren't automatically derived can simply implement them if desired:
(https://doc.rust-lang.org/nomicon/send-and-sync.html)

FWIW, I don't need to actually send namespaces. But I need `ConstraintSystem` to be `Send` so I can send `ProvingAssignment` (which responds affirmatively to `is_extensible()`) but *not* other `ConstraintSystem`s (like `Namespace`).

If you know a better way to make it work, I'm open. But can you explain why the quotation I included above doesn't apply here?